### ISSUE TITLE
Data elaboration creates a black band

### ISSUE THREAD

**Issue by NoxFelis** [matciotola/hyperspectral_pansharpening_toolbox#4]
Hello, 
I am currently using your toolbox and when applying different methods on the PRISMA data, I noticed the ERGAS metric pulled out a "nan". Turns out there was a division by 0 and when trailing up to find the source of this issue, I found that during the "elaborate" step, the 56th band of the test ground truth images comes out as entirely black. Zeros everywhere. I am still trying to figure out how this happened as it makes little sense to me.

**Comment by NoxFelis** [comment]
Hello again
after further inspection I have found what might cause the issue.
in the function he5_to_mat the VNIR cube has 66 bands
The last 4 bands are zeros, most are taken care of with the list_bands variable
except the 63rd band which is kept. 
``` LIST_BANDS = np.concatenate([list(range(7, 63)), list(range(71, 86)), list(range(89, 107)), list(range(120, 150)), list(range(176, 216))])```
but if I remove it the hs cube will have 158 bands in total instead of 159. Is it possible to know which other band has then been missed? Thank you

**Comment by matciotola** [comment]
Hi @NoxFelis,

Thank you for your message. Could you clarify if this issue is occurring with all the images or only with a specific one? If it’s just one, could you please let me know which image?

Thanks,
Matteo

**Comment by NoxFelis** [comment]
Hello,
For now I have only looked into the Test dataset. Two of the images seem to have this issue: Cagliari and Udine. 
When you upload the VNIR Cube of 66 bands, the last 4 are zeros. Keeping the 63rd band is keeping one black band (see the two spectral profiles of the GT simulated images). 
Considering the algorithms have been trained for the specific bands you selected (see list_bands above) I have decided to remove said problem bands when applying the reduced resolution metrics.

I have added the following bit of code at the beginning of each metric function : 
```mask = (labels == 0)
band_all_zero = mask.all(dim=(0,2,3))
y = torch.where(band_all_zero)[0]
for i in range(len(y)):
    outputs = torch.cat((outputs[:,:y[i]], outputs[:,y[i]+1:]), axis = 1)
    labels = torch.cat((labels[:,:y[i]], labels[:,y[i]+1:]), axis = 1)
    y[i] = y[i]-1
```

Thank you again for all this work,

<img width="738" height="343" alt="Image" src="https://github.com/user-attachments/assets/c8f64486-e08b-4678-b3cd-f2642c8bad8c" />
<img width="740" height="345" alt="Image" src="https://github.com/user-attachments/assets/5be62681-d5e4-487f-ae4b-da732a4f10db" />

**Comment by matciotola** [comment]
Thank you @NoxFelis for your answer and explanation. 
I have double-checked both images and their values. As you mentioned, from bands 63 to 66 the values are indeed zero. However, in our code we only consider up to band 62 (inclusive). Specifically:
`list(range(7, 63))` 
which generates values from 7 to 62 (56 bands). Could you please confirm this behavior?
In the meantime, I am re-downloading the data from the PRISMA portal to verify whether anything has changed compared to our original dataset.
Would it also be possible for you to check if this same behavior occurs with the Full Resolution data?
Thanks a lot,
Matteo

**Comment by NoxFelis** [comment]
Hello again,
I am sorry I may have expressed myself wrongly. I'll show you a screenshot of when debugging.

This is when opening the VNIR Cube
<img width="1919" height="936" alt="Image" src="https://github.com/user-attachments/assets/02f3e79c-e341-4e4f-9926-b77a24755659" />

This is when selecting the bands after concatenating the cubes. The band 62 becomes band 55 (last band of the filtered VNIR Cube)

<img width="1920" height="935" alt="Image" src="https://github.com/user-attachments/assets/5c57073d-58eb-4159-a0e8-37c0d117e246" />

I can check for the Full Resolution Data but I might reply for that in a few days if that's ok?

**Comment by matciotola** [comment]
Dear @NoxFelis,
I have re-downloaded both images from the PRISMA portal and checked them carefully.
With both the old and new datasets, I did not find any blank bands. Unfortunately, I am not able to reproduce the issue on my side, and I am really sorry about that.
Please feel free to contact me by email (matteo.ciotola@unina.it) so that we can work together to find a suitable workaround for your problem.
For reference, I leave here the code I used to verify the filtered dataset (this was my final check, after analyzing the hyperspectral stack band by band).
Thank you,
Matteo

```
import numpy as np
import h5py

list_bands = np.concatenate([list(range(7, 63)), list(range(71, 86)), list(range(89, 107)), list(range(120, 150)), list(range(176, 216))])


filename = "/Users/matteo/Downloads/PRS_L2D_STD_20230824100356_20230824100400_0001.he5"
filename2 = "/Users/matteo/Downloads/PRS_L2D_STD_20220905101901_20220905101905_0001.he5"

with h5py.File(filename, 'r') as h5f:
    SWIR = np.array(h5f['HDFEOS/SWATHS/PRS_L2D_HCO/Data Fields/SWIR_Cube'][()], dtype=np.uint16)
    VNIR = np.array(h5f['HDFEOS/SWATHS/PRS_L2D_HCO/Data Fields/VNIR_Cube'][()], dtype=np.uint16)
    pan = np.array(h5f['HDFEOS/SWATHS/PRS_L2D_PCO/Data Fields/Cube'][()], dtype=np.uint16)

    VNIR = np.flip(np.moveaxis(VNIR, 1, 2), axis=-1)
    SWIR = np.flip(np.moveaxis(SWIR, 1, 2), axis=-1)

    hs = np.concatenate((VNIR, SWIR), axis=2)
    hs_filtered = hs[:, :, list_bands]


with h5py.File(filename2, 'r') as h5f:
    SWIR2 = np.array(h5f['HDFEOS/SWATHS/PRS_L2D_HCO/Data Fields/SWIR_Cube'][()], dtype=np.uint16)
    VNIR2 = np.array(h5f['HDFEOS/SWATHS/PRS_L2D_HCO/Data Fields/VNIR_Cube'][()], dtype=np.uint16)
    pan2 = np.array(h5f['HDFEOS/SWATHS/PRS_L2D_PCO/Data Fields/Cube'][()], dtype=np.uint16)

    VNIR2 = np.flip(np.moveaxis(VNIR2, 1, 2), axis=-1)
    SWIR2 = np.flip(np.moveaxis(SWIR2, 1, 2), axis=-1)

    hs2 = np.concatenate((VNIR2, SWIR2), axis=2)
    hs2_filtered = hs2[:, :, list_bands]


for i in range(hs_filtered.shape[-1]):
    checked_band = hs_filtered[:, :, I]
    if (checked_band == 0).all():
        print(i)

for i in range(hs2_filtered.shape[-1]):
    checked_band = hs2_filtered[:, :, I]
    if (checked_band == 0).all():
        print(i)
```

**Comment by NoxFelis** [comment]
Hello @matciotola,
I just wanted to say I have redownloaded the two images, and checked with your code and in ENVI, and indeed, the band 62 is not black. I suppose there has been an issue on my side when downloading the PRISMA data or unzipping the files. 
Thank you again for the time taken to double check. 
Best regards,
Nox

**Comment by matciotola** [comment]
@NoxFelis I’m really glad to hear that you solved the problem. That’s the most important thing!
If you run into any other issues, please don’t hesitate to reach out to us.
Thank you,
Matteo
